### PR TITLE
fix: return NaN for invalid cluster counts in CalinskiHarabaszScore

### PR DESCRIPTION
**Pull request by nightcityblade**
Related to #3655

Description:

`CalinskiHarabaszScore` crashes with a `ValueError` when all samples belong to a single cluster, because `sklearn.metrics.calinski_harabasz_score` requires at least 2 unique labels. This is the same issue that was fixed for `SilhouetteScore` and `DaviesBouldinScore` in #3654.

This PR adds a guard that returns `float("nan")` when the number of unique labels is less than 2, preventing the engine from crashing during early training epochs.

Changes:
- Added cluster count validation in `_calinski_harabasz_score()` 
- Added docstring note about NaN return behavior
- Added `test_single_cluster_returns_nan` test

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: ignite/metrics/clustering/calinski_harabasz_score.py [hunk 1]
```diff
@@ -14,6 +14,11 @@ def _calinski_harabasz_score(features: Tensor, labels: Tensor) -> float:
 
     np_features = features.cpu().numpy()
     np_labels = labels.cpu().numpy()
+
+    n_unique_labels = len(set(np_labels))
+    if n_unique_labels < 2:
```
Comment (TahaZahid05) [inline ignite/metrics/clustering/calinski_harabasz_score.py]: Validation logic should also cover the case `n_unique_labels == n_samples`. We should also display a warning to the user. Let's also add a `try...except` block so if we do get any future sklearn errors, it doesn't break the engine

File: tests/ignite/metrics/clustering/test_calinski_harabasz_score.py [hunk 2]
```diff
@@ -19,6 +19,18 @@ def test_zero_sample():
...
+    """CalinskiHarabaszScore should return NaN when all samples belong to a single cluster."""
+    import math
+
+    metric = CalinskiHarabaszScore()
+    features = torch.randn(10, 5)
+    labels = torch.zeros(10, dtype=torch.long)
+    metric.update((features, labels))
+    result = metric.compute()
+    assert math.isnan(result)
+
```
Comment (TahaZahid05) [inline tests/ignite/metrics/clustering/test_calinski_harabasz_score.py:30]: Add another test case for the validation check `n_unique_labels == n_samples`

File: ignite/metrics/clustering/calinski_harabasz_score.py (hunk 1 above)
Comment (aaishwarymishra) [reply ignite/metrics/clustering/calinski_harabasz_score.py]: I don't think we should add exception handling with try except, if there is a problem the engine should throw error and stop.

Comment (TahaZahid05) [reply ignite/metrics/clustering/calinski_harabasz_score.py]: True. we can just let sklearn errors surface.

**Comment by nightcityblade** [comment]
Thanks for the review! I've updated to also handle the case where `n_unique_labels == n_samples` (each sample is its own cluster), returning NaN. Added a corresponding test case `test_each_sample_own_cluster_returns_nan` as well.

File: ignite/metrics/clustering/calinski_harabasz_score.py [hunk 3]
```diff
@@ -14,6 +14,12 @@ def _calinski_harabasz_score(features: Tensor, labels: Tensor) -> float:
 
     np_features = features.cpu().numpy()
     np_labels = labels.cpu().numpy()
+
+    n_unique_labels = len(set(np_labels))
+    n_samples = len(np_labels)
+    if n_unique_labels < 2 or n_unique_labels == n_samples:
```
Comment (TahaZahid05) [inline ignite/metrics/clustering/calinski_harabasz_score.py]: Looking at the other PR patching the other two metrics, better to change `==` to `>=` as it is more defensive:

```suggestion
    if n_unique_labels < 2 or n_unique_labels >= n_samples:
```

**Review by TahaZahid05 — COMMENTED** [review]
Rest looks fine, just change the check from `==` to `>=`

**Comment by nightcityblade** [comment]
Thanks @TahaZahid05! Changed `==` to `>=` for a more defensive check, as suggested.

**Comment by nightcityblade** [comment]
Thanks for the approval @TahaZahid05! 🎉

File: ignite/metrics/clustering/calinski_harabasz_score.py [hunk 4]
```diff
@@ -14,6 +14,12 @@ def _calinski_harabasz_score(features: Tensor, labels: Tensor) -> float:
 
     np_features = features.cpu().numpy()
     np_labels = labels.cpu().numpy()
+
+    n_unique_labels = len(set(np_labels))
```
Comment (vfdev-5) [inline ignite/metrics/clustering/calinski_harabasz_score.py]: Can't we use numpy methods to get the number of unique values instead of using python set ?

File: ignite/metrics/clustering/calinski_harabasz_score.py (hunk 3 above)
Comment (vfdev-5) [reply ignite/metrics/clustering/calinski_harabasz_score.py]: Maybe, it does cover more, but looks a bit weird :)
OK, let's keep it like this...

File: tests/ignite/metrics/clustering/test_calinski_harabasz_score.py [hunk 5]
```diff
@@ -19,6 +19,30 @@ def test_zero_sample():
         metric.compute()
 
 
+def test_single_cluster_returns_nan():
```
Comment (vfdev-5) [inline tests/ignite/metrics/clustering/test_calinski_harabasz_score.py]: Let's unify the test cases using pytest.mark.parametrize

**Comment by nightcityblade** [comment]
Thanks @vfdev-5 for the review! I've addressed all your feedback:

1. ✅ Replaced `set()` with `np.unique()` for label counting
2. ✅ Kept `>=` for the defensive check
3. ✅ Unified the NaN test cases using `pytest.mark.parametrize`

Force-pushed with the changes.

**Comment by vfdev-5** [comment]
@nightcityblade seems like you messed up with the git commits history.

**Comment by nightcityblade** [comment]
Sorry about the messy commit history @vfdev-5! I've force-pushed a clean single commit rebased on master. The change is the same — adds NaN return for invalid cluster counts with tests. Let me know if anything else needs fixing.

File: ignite/metrics/clustering/calinski_harabasz_score.py [hunk 6]
```diff
@@ -14,6 +16,12 @@ def _calinski_harabasz_score(features: Tensor, labels: Tensor) -> float:
 
     np_features = features.cpu().numpy()
     np_labels = labels.cpu().numpy()
+
+    n_unique_labels = len(np.unique(np_labels))
```
Comment (TahaZahid05) [inline ignite/metrics/clustering/calinski_harabasz_score.py]: I think using `torch.unique` is preferred to avoid switching to cpu.

File: ignite/metrics/clustering/calinski_harabasz_score.py [hunk 7]
```diff
@@ -14,6 +16,12 @@ def _calinski_harabasz_score(features: Tensor, labels: Tensor) -> float:
 
     np_features = features.cpu().numpy()
     np_labels = labels.cpu().numpy()
```
Comment (TahaZahid05) [inline ignite/metrics/clustering/calinski_harabasz_score.py]: add `.detach()` to avoid runtime issues.

File: ignite/metrics/clustering/calinski_harabasz_score.py [hunk 8]
```diff
@@ -1,6 +1,8 @@
 from collections.abc import Callable
 from typing import Any
 
+import numpy as np
```
Comment (TahaZahid05) [inline ignite/metrics/clustering/calinski_harabasz_score.py]: we can remove the import as well if we use `torch.unique`

File: tests/ignite/metrics/clustering/test_calinski_harabasz_score.py [hunk 9]
```diff
@@ -11,6 +11,24 @@
...
 
+
+@pytest.mark.parametrize(
+    "features, labels, desc",
+    [
+        (torch.randn(10, 5), torch.zeros(10, dtype=torch.long), "single cluster"),
+        (torch.randn(5, 3), torch.arange(5, dtype=torch.long), "each sample own cluster"),
+    ],
+)
+def test_invalid_cluster_count_returns_nan(features, labels, desc):
```
Comment (TahaZahid05) [inline tests/ignite/metrics/clustering/test_calinski_harabasz_score.py]: We can test on both `True` and `False` for `check_compute_fn` now.

**Comment by nightcityblade** [comment]
Thanks for the detailed feedback! I've addressed all review comments:

- Replaced `np.unique` with `torch.unique(labels.detach())` to avoid unnecessary CPU transfer
- Removed the now-unused `numpy` import
- Added `check_compute_fn` parametrization (`True`/`False`) to the invalid cluster count tests

Let me know if anything else needs adjusting.

**Comment by TahaZahid05** [comment]
I don't see the commit. kindly fix it.

**Comment by nightcityblade** [comment]
Force-pushed a clean rebase on latest master. The commit should be visible now @TahaZahid05 — single commit with the NaN validation using `torch.unique(labels.detach())`. Tests pass locally.

**Comment by vfdev-5** [comment]
Code style check is failing: https://github.com/pytorch/ignite/actions/runs/23405930582/job/68113360470?pr=3656

**Comment by nightcityblade** [comment]
Fixed the black formatting issue — removed the extra blank line in the test file. CI should pass now. Thanks for the review!

**Comment by nightcityblade** [comment]
Thanks for the reviews @TahaZahid05 @vfdev-5! I've addressed the feedback:
- Added `check_compute_fn` parametrize (`True`/`False`) to the invalid cluster count test
- The code already uses `torch.unique` with `.detach()` and `>=` for the bounds check

Ready for re-review!
